Bug 1875465 - Add tab strip for tablets

rahulsainani opened the pull request:
### What
– Add Secret Settiings - "Enable Tab Strip" that's visible only on tablets.
– Display open tabs in a Tab Strip above toolbar
– Add Ability to add a new tab and close tabs.

While these are not features of the Tab Strip UI, these are changes are in Settings when tab strip is enabled:
– toolbar customization in Settings is not shown as Toolbar defaults to Top.
– Swipe tab gesture to switch tabs in settings is not shown as it is disabled.


### How
– Update toolbar position based on form factor. This required replacing usages of `settings.shouldUseBottomToolbar` to `settings.toolbarPosition`. 
– `ToolbarPosition` is defaults to `TOP` when it's a tablet. This is done to simplify the implementation.
– Add dimensions to help update layouts and positioning logic
– Display TabStrip on `HomeFragment` and `BrowserFragment`. It would be ideal to join the two fragments but that's outside the scope right now.
– Add `TabStrip` composable that provides callbacks for adding a tab, selecting a tab and removal of the last tab. These are used by `BrowserFragment` and `HomeFragment` differently. 
– Create `TabUIState` model that's mapped from `TabSessionState`. This simplifies the observation and the only contains information required to render the `TabStrip`. Also helps in recompositions since only if the content in this model changes would it recompose.
– LazyRow to show `Tabs`

### Preview
[tablet-tab-strip-3.webm](https://github.com/mozilla-mobile/firefox-android/assets/38040960/fd50937d-5442-494e-b4aa-0baf75569a57)



### Known issues
–~~There's ghost padding at the bottom of `EngineView`. Keener eyes than mine could help solve this.~~

### Non exhastive list of Todos
- [x] Add secret settings feature flag
- [x] Remove customize toolbar setting for tablets
- [x] Update scroll when selecting tabs items that are partially visible.
- [ ] `TabStripState` can be marked `@Immutable` to make compose compiler aware so it doesn't need to recompose when the list doesn't change. This can also be solved by using Immutable collections from kotlinx.
- [x] Same tab width for all tabs based on number of tabs.
- [x] Display url when `content.title` is blank.
- [x] Hide tab strip when full screen.
- [ ] Tab strip should scroll up when toolbar scrolls up with content.


### Note for reviewers
This isn't pulling from a specific design but is using the usual design system tokens. 


### 26 Feb Update 
Any remaining bugs and functionality will be tracked separate as part of this [meta](https://bugzilla.mozilla.org/show_bug.cgi?id=1882105). 


------------------------------------------------------------------------------------------------------------------------

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.









































### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1875465

Kuuh4 commented:
Very Nice! Congratulations and thanks!

riddlerboss248 commented:
Would love to have it on mobile too. Like how brave(ios), Vivaldi and opera have it.

NVins commented:
I support! please support tab strip for tablets, for the mobile version of Firefox Android.

NVins commented:
Please also add a tablet tab bar for the mobile version of Firefox Android.

danieljl commented:
Thanks! Please also consider adding an option to use the old (current) tab UI for users who prefer it on their smaller tablets.

sukhrob-shakirov commented:
Finally!! thanks

tkompis commented:
Hello! Is it possible to download an APK to test/try this feature?

tkompis commented:
A recommendation that popped in my mind, is the addition of a tab width limiter/lock. Depending on the length of a website's name, all tabs are different lengths on the tab bar.

fabiocarballo commented:
soo good! need this shipped!

Nick768 commented:
Ok, I've tested it and I have to say, it works! I know it crashes sometimes, but this could also be because it is built with an alpha level gecko version. The only problems so far:

1. Some Tabs doesn't show titles (I think every github tab)
2. The blank area at the bottom (it has exactly the same size as the normal FF navigation bar, therefore shouldn't be hard to find when searching for references to it)
3. In fullscreen mode (e.g. YouTube fullscreen playback) the tabbar is still there. That shouldn't be of course
4. It should be done quick because I need it so much but don't have time to contribute :-)

But after all this is what every tablet user is waiting for. Since it isn't production ready I have to stick with Firefox in Termux with some other drawbacks...

Nick768 commented:
Thank you, you are amazing! Everything is fixed, except YouTube fullscreen, but my Tablet has a 16:10 screen, therefore a 16:9 video fits even with the tab bar shown...

girst commented:
Thank you for working on this!

Just looking at  tablet-tab-strip.webm, it feels very large. On desktop the tall tabs don't take up much space relative to the screen, but on smaller tablets every pixel counts. I hope you are considering an option to make them smaller, like `browser.uidensity;1` on desktop.

Aasikki commented:
> Thank you for working on this!
> 
> Just looking at tablet-tab-strip.webm, it feels very large. On desktop the tall tabs don't take up much space relative to the screen, but on smaller tablets every pixel counts. I hope you are considering an option to make them smaller, like `browser.uidensity;1` on desktop.

Yeah this is awesome otherwise but the tabs are absolutely huge even on a tab with 14.6" display.

rahulsainani commented:
Thanks everyone for already trying and testing this version and the feedback, it's really appreciated! 🙌 I will try to summarize some of the updates and address some of the questions in this message. The video now shows the updated version and shows that the old/current tabs tray, that is not going anywhere. 

– @tkompis tab width should be the same now. This PR is on top of last week's main branch so all the other functionalities should work if you want to set it as your default. However, you would miss out on the other improvements that keep happening on the main branch, till this is merged. 
– @Nick768 full screen issue should also be solved now too. Could you please share when/how the app crashed, that really shouldn't be happening. The shortcuts are a great idea, we've talked about that, but that won't be part of this PR. 
– @girst @Aasikki Valid points! Could you please check if the tabs are still large on the latest update? If so, could you please share the device details and possiblly a link to the screenshot?

Thanks again!

tkompis commented:
> Thanks everyone for already trying and testing this version and the feedback, it's really appreciated! 🙌 I will try to summarize some of the updates and address some of the questions in this message. The video now shows the updated version and shows that the old/current tabs tray, that is not going anywhere.
> 
> – @tkompis tab width should be the same now. This PR is on top of last week's main branch so all the other functionalities should work if you want to set it as your default. However, you would miss out on the other improvements that keep happening on the main branch, till this is merged. – @Nick768 full screen issue should also be solved now too. Could you please share when/how the app crashed, that really shouldn't be happening. The shortcuts are a great idea, we've talked about that, but that won't be part of this PR. – @girst @Aasikki Valid points! Could you please check if the tabs are still large on the latest update? If so, could you please share the device details and possiblly a link to the screenshot?
> 
> Thanks again!

The updated video is showing as corrupt, I can't see it.

I've decided to be patient and wait for it to be merged into the main branch other than using it right now. Still, superb work, and it shows why Firefox is the best browser. I wish I was more knowledgeable in terms of programming, to help!

Aasikki commented:
> Thanks everyone for already trying and testing this version and the feedback, it's really appreciated! 🙌 I will try to summarize some of the updates and address some of the questions in this message. The video now shows the updated version and shows that the old/current tabs tray, that is not going anywhere.
> 
> – @tkompis tab width should be the same now. This PR is on top of last week's main branch so all the other functionalities should work if you want to set it as your default. However, you would miss out on the other improvements that keep happening on the main branch, till this is merged. – @Nick768 full screen issue should also be solved now too. Could you please share when/how the app crashed, that really shouldn't be happening. The shortcuts are a great idea, we've talked about that, but that won't be part of this PR. – @girst @Aasikki Valid points! Could you please check if the tabs are still large on the latest update? If so, could you please share the device details and possiblly a link to the screenshot?
> 
> Thanks again!

Yeah the tabs still seem to be the same size.
![Screenshot_20240219_132733_Firefox Fenix](https://github.com/mozilla-mobile/firefox-android/assets/58399890/6d27bfc6-3771-4192-8265-067650e8dc6c)

Here's chrome on the same tab for comparison.
![Screenshot_20240219_132726_Chrome](https://github.com/mozilla-mobile/firefox-android/assets/58399890/cf0bd639-5bda-4abf-8835-cf3a6619f8f4)

This is on a Samsung Galaxy Tab S9 Ultra.
I also noticed that there probably should be an animation when making  a new tab.
Huge thinks for working on this, can't wait to also switch to Firefox on the tab!

rahulsainani commented:
@tkompis Refreshing the page should work. I've updated the video again, maybe now you can see it. Your help in testing and feedback has been great!

@Aasikki Thanks for sharing the screenshots! The tabs should adjust their width when there are more tabs. I've reduced the  height and updated the paddings so it looks a bit more clean. The new tab animation might be outside the scope of this change since we open the home page on clicking new tab.

Aasikki commented:
> @tkompis Refreshing the page should work. I've updated the video again, maybe now you can see it. Your help in testing and feedback has been great!
> 
> @Aasikki Thanks for sharing the screenshots! The tabs should adjust their width when there are more tabs. I've reduced the height and updated the paddings so it looks a bit more clean. The new tab animation might be outside the scope of this change since we open the home page on clicking new tab.

Definitely better!
I think it could be just a bit thinner still (a quick photoshop I made):
![Slimmer copy](https://github.com/mozilla-mobile/firefox-android/assets/58399890/36ea5c85-b09d-43c9-9877-10b8d68accb6)

Maybe it should also have a darker background to appear more like Firefox on desktop. Looks a bit sleeker that way imo.
![Slimmer and darker copy](https://github.com/mozilla-mobile/firefox-android/assets/58399890/5cf9b85e-b9bd-4cb0-9587-0bf41c6696a6)

Nothing wrong with the width of the tabs, but I noticed that scroll to hide toolbar doesn't seem to be working for me.

Nick768 commented:
Checked your last version and now everything seems to work! Thanks!

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+}
+
+@Composable
+private fun TabStripContent(
+    state: TabsStripState,
+    onAddTabClick: () -> Unit,
+    onCloseTabClick: (id: String) -> Unit,
+    onSelectedTabClick: (id: String) -> Unit,
+) {
```
one downside to the composable + content pattern is that the previews will no longer test the store observations. Have you considered just passing in `BrowserStore` and `AppStore` as explicit parameters instead?

The `tabsUseCases` could be hoisted as well and be incorporated into click lambdas. From a cursory look, the `tabsUseCases` might be able to be constructed within previews, since their only dependency is `BrowserStore`, so we could have fully functional previews using our stores + use cases without needing to use `components` at this level!

rahulsainani replied:
Yeah I did consider this and initially had it like that but that was leading to duplication of calls to tabsUsecases functions in both the places TabStrip is used. This `TabStrip` top level composable acts like a controller/fragment and the `Content` composable is the UI that simply gets the data. So the parent fragments don't have to setup similarly everywhere. We can extract another top level composable that uses components and this composable (can be made private) can simply observe but I'm not sure if there's benefit to testing observation in previews since the mapper is tested and observation code is tested, and the content composable gets the simplified data model.

MozillaNoah replied:
I just investigated a potential option for this that combines both approaches:

```
@Composable
fun TabsStrip(
    appStore: AppStore = components.appStore,
    browserStore: BrowserStore = components.core.store,
    tabsUseCases: TabsUseCases = components.useCases.tabsUseCases,
    onHome: Boolean = false,
    onAddTabClick: () -> Unit,
    onLastTabClose: () -> Unit = {},
    onSelectedTabClick: () -> Unit = {},
)
```
```
@Composable
@Preview
private fun TabStripPreview() {
    val browserStore = BrowserStore()

    FirefoxTheme {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            contentAlignment = Alignment.Center,
        ) {
            TabsStrip(
                appStore = AppStore(),
                browserStore = browserStore,
                tabsUseCases = TabsUseCases(browserStore),
                onAddTabClick = {
                    val tab = createTab(
                        url = "www.example.com",
                    )
                    browserStore.dispatch(TabListAction.AddTabAction(tab))
                },
                onLastTabClose = {},
                onSelectedTabClick = {},
            )
        }
    }
}
```

Giving us:

https://github.com/mozilla-mobile/firefox-android/assets/87384386/8b6012c9-8c60-4568-9c66-7eeff99ee691

MozillaNoah replied:
And we could still keep the existing `content` composable and the parameterized previews if we wanted.

rahulsainani replied:
Thanks for writing down your thoughts and the video! And the setup cost isn't much really too. That wasn't the best reasoning from my side. After thinking more, I agree this is much better. I've updated and removed the default params for functions, they should be explicit too.

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+        IconButton(onClick = onAddTabClick) {
+            Icon(
+                painter = painterResource(R.drawable.mozac_ic_plus_24),
+                modifier = Modifier.size(20.dp),
```
is the `20` to align the close buttons with the open tab button? If so, we could extract this into a constant!

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+) {
+    BoxWithConstraints(modifier = modifier) {
+        val listState = rememberLazyListState()
+        val tabWidth = ((maxWidth - tabStripStartPadding) / state.tabs.size) - spaceBetweenTabs
```
could we get a small comment here explaining the reasoning behind the math for future maintainers?

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+                        ),
+                )
+
+                Spacer(modifier = Modifier.size(spaceBetweenTabs))
```
alternatively, we could use:
```
horizontalArrangement = Arrangement.spacedBy(spaceBetweenTabs),
```
inside of `LazyRow`

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+    TabStripCard(
+        modifier = modifier.fillMaxSize(),
+        backgroundColor =
+        if (state.isPrivate) {
+            if (state.isSelected) {
+                FirefoxTheme.colors.layer3
+            } else {
+                FirefoxTheme.colors.layer2
+            }
+        } else {
+            if (state.isSelected) {
+                FirefoxTheme.colors.layer2
+            } else {
+                FirefoxTheme.colors.layer3
+            }
+        },
```
a `when` might help visually clean this up

t-p-white replied:
could even extract this block of logic to a func altogether, such as `getBackgroudColor`

rahulsainani replied:
I think if reads better here since there are two nested conditions, I will wait to get the final background colors before extracting it out in a function, think the updated colors might just simplify the logic.

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+            IconButton(onClick = { onCloseTabClick(state.id) }) {
+                Icon(
+                    painter = painterResource(R.drawable.mozac_ic_cross_20),
+                    modifier = Modifier.size(20.dp),
```
you shouldn't need `Modifier.size` with a `_20` vector

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+                .clip(CircleShape),
+        )
+    } else {
+        Icon(
+            painter = painterResource(R.drawable.mozac_ic_globe_24),
+            modifier = Modifier.size(tabStripIconSize),
+            tint = FirefoxTheme.colors.iconPrimary,
+            contentDescription = null,
+        )
```
if we're wanting to be consistent, [we're using a FavIcon as the fallback in the Tabs Tray](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/compose/TabThumbnail.kt#74-90)

rahulsainani replied:
Nice! Thanks! I wanted to look into this, also now we don't need the bitmap icon

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:
```diff
@@ -138,6 +140,14 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
         }
     }
 
+    private fun setupTabStripPreference() {
```
this doesn't necessarily need to be in its own one-off function

rahulsainani replied:
Shaped by detekt, so we don't add detekt suppression here. Think every item here can be a one off function to prevent it further but that's outside the scope of this PR.

MozillaNoah reviewed (commented):
this has been a long time coming and I can see the work you've already put into it! This is a pretty substantial PR, so please pardon the volume of comments as a result.

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/utils/Settings.kt:
```diff
@@ -854,6 +854,9 @@ class Settings(private val appContext: Context) : PreferencesHolder {
             return touchExplorationIsEnabled || switchServiceIsEnabled
         }
 
+    private val isTablet: Boolean
+        get() = appContext.resources.getBoolean(R.bool.tablet)
```
hmm, interesting. we'll use the same values-sw600dp to know if it's a tablet for the grid system as well. JFYI this is an existing boolean that we already use to detect if it's tablet across the codebase and in general that's the way.

MozillaNoah replied:
I know it's pre-existing; was just kinda musing aloud that there may be a better way of handling this, longer term, without relying upon XML resources

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/home/ToolbarView.kt:
```diff
@@ -66,16 +66,27 @@ class ToolbarView(
                     gravity = Gravity.TOP
                 }
 
+                val isTabletAndTabStripEnabled = context.resources.getBoolean(R.bool.tablet)
```
I don't thinkg we need another feature flag for it other than secret settings, I'll think more about it going forward but the whole idea is to have this feature enabled

rahulsainani commented on fenix/app/src/main/res/values-sw600dp/dimens.xml:
```diff
@@ -0,0 +1,7 @@
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<resources>
+    <dimen name="tabs_strip_height">48dp</dimen>
```
We need this so on phones, this is 0dp. This needs to be set in xml layouts so the size is set by the parent. We don't want to set the height in compose (which is currently defining the UI but the parent is still View which has the constraints, margins and the size)

MozillaNoah replied:
won't it still be marked as `GONE` though because of the logic with the `tablet` boolean dimen?

rahulsainani replied:
You're right there. However, moving forward, as we enable it by default the check in many places will change from `isTabletAndTabStripEnabled` to `isTablet`. Now `isTablet` is simply sw600dp check. We can simplify that by having 2 `tabs_strip_height` dimensions for phones and tablets. Simplifying the marging upate calls to one lines instead of if (tablet ) 48 else 0.

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -654,6 +658,30 @@ class HomeFragment : Fragment() {
+                    TabsStrip(
+                        onHome = true,
+                        onAddTabClick = {
+                            sessionControlInteractor.onNavigateSearch()
+                        },
+                        onSelectedTabClick = {
+                            (requireActivity() as HomeActivity).openToBrowser(BrowserDirection.FromHome)
```
When we select a tab, we want to open a browser and see the tab 😄 what's the alternate?

MozillaNoah replied:
I'm not saying there _could_ be an alternate; I was moreso just trying to check-in if UX has been looped in with this feature potentially landing soon.

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,371 @@
+fun TabsStrip(
+    onHome: Boolean = false,
+    onAddTabClick: () -> Unit,
+    onLastTabClose: () -> Unit = {},
+    onSelectedTabClick: () -> Unit = {},
```
TabStrip is used in two places - Home and browser fragment, and these two are only needed in one of those places.

MozillaNoah replied:
I guess it's a matter of lambda implementation enforcement. Home and Browser could both implement the empty lambdas themselves, but then `TabStrip` wouldn't have to care about whether they were correctly implemented during integration. 

There's not necessarily right answer in this situation.

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt:
```diff
@@ -0,0 +1,369 @@
+            .clip(CircleShape),
+        contentAlignment = Alignment.Center,
+    ) {
+        Favicon(
```
I hope my last comment with this block wasn't misleading. We should still use the `icon` bitmap from `TabSessionState.content` here, if it's not null. Otherwise, `Favicon` will be doing a network call or making a creating custom asset for _every_ tab icon.

MozillaNoah replied:
https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/compose/TabThumbnail.kt#74-91

rahulsainani replied:
I did do that but then realised that Favicon does cache icons in memory, similar to how `ContentState.icon` is in memory, so we don't have to worry about it making multiple requests. Or am I misunderstanding something here?

MozillaNoah replied:
OK so I had to double check. Favicon does indeed pull from the same memory cache that drives `ContentState.icon` (via `BrowserIcons`). 

I suppose it would come down to some theoretical trade offs of:
- Using the bitmap that's already out of cache and inside of `ContentState.icon`
- Is additional loading from memory within `BrowserIcons` trivial? [source code](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt#291-297)

I don't have the scientific answers to those, but if the bitmap is already inside of `ContentState.icon`, we shouldn't have to pull it out of the cache again.

rahulsainani replied:
I think it's fine in this case as the cache (LRUCache) is simply holding the bitmap, and the logic is simply getting the Bitmap, there's no decoding from bytes happening. And It simplifies the UI even further.

MozillaNoah replied:
OK, I'm fine with moving forward with this.

MozillaNoah commented:
I'm seeing like a 2dp space between the toolbar and the bottom of the screen, when the toolbar is set to the bottom
![image](https://github.com/mozilla-mobile/firefox-android/assets/87384386/c494a8b3-ad12-4194-92a6-92b4b21745f8)

regs01 commented:
In light theme active tab barely distinguishable. Needs some shadow and border around to highlight it.

Like on desktop
![](https://github.com/mozilla-mobile/firefox-android/assets/10468217/a74060d1-57b6-4397-ac86-915567fabc38)

rahulsainani commented:
> I'm seeing like a 2dp space between the toolbar and the bottom of the screen, when the toolbar is set to the bottom.
@MozillaNoah that should be fixed now.

Kuuh4 commented:
Omg it's been a month and so much happened! Can't wait to test it [I'm going to look how to do it right now]. Thanks for the work! Wish we could see vertical tabs too, would be awesome. [both expanding with mouse hover, and with toggle button for touch at the top of the retracted strip with just the icons, quite a bit like edge does it very well ]

MozillaNoah commented on fenix/app/src/main/java/org/mozilla/fenix/settings/CustomizationFragment.kt:
```diff
@@ -140,7 +151,7 @@ class CustomizationFragment : PreferenceFragmentCompat() {
     }
 
-    private fun setupGesturesCategory() {
+    private fun setupGesturesCategory(isSwipeToolbarToSwitchTabsVisible: Boolean) {
```
I initially missed this new boolean. maybe add a comment that the swipe gesture should not be enabled when the tab strip is visible?

IrradiatedHaggis commented:
This looks incredible!  Thank you so much.  Are there currently plans to include keyboard shortcuts?  Currently Mozilla Firefox on Android is the only browser I've used with my Galaxy Tab S7+ which doesn't respond to shortcuts like Control+T.

MozillaNoah commented:
> @MozillaNoah that should be fixed now.

Confirmed I no longer see the issue.

Verdi commented:
From a UX perspective, this looks good enough to land behind a pref in nightly.

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt:
```diff
@@ -536,7 +536,9 @@ class DefaultTabsTrayController(
                 TabsTray.openedExistingTab.record(TabsTray.OpenedExistingTabExtra(source ?: "unknown"))
                 tabsUseCases.selectTab(tab.id)
-                browsingModeManager.mode = BrowsingMode.fromBoolean(tab.content.private)
+                val mode = BrowsingMode.fromBoolean(tab.content.private)
+                browsingModeManager.mode = mode
+                appStore.dispatch(AppAction.ModeChange(mode))
```
Fair point! as discussed offline, basically we need the AppStore to be notified of the mode change from tabs tray so tab strip (that is observing the AppStore for browsing mode) can react to changes. Here's the [PR](https://github.com/mozilla-mobile/firefox-android/pull/5733).

csolisr commented:
Talking about flags - is there one to fake my relatively large phone into counting as a tablet for interface purposes?

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/search/toolbar/ToolbarView.kt:
```diff
@@ -123,6 +125,12 @@ class ToolbarView(
+
+            if (settings.isTabletAndTabStripEnabled) {
+                (layoutParams as ViewGroup.MarginLayoutParams).updateMargins(
+                    top = context.resources.getDimensionPixelSize(R.dimen.tabs_strip_height),
```
Continuing from this [comment](https://github.com/mozilla-mobile/firefox-android/pull/5218#discussion_r1502426884), we want to use the same value as `tabs_strip_height` here. Don't think we need a new dimension that has the same value or simply references tabs_strip_height internally.

MozillaNoah replied:
I guess I'm just thinking forward to when UX may ask to have distinct values of margin vs height. This is by no means a blocker and maybe something to simply keep in mind as this feature develops and we eventually get some official specs.

rahulsainani commented on fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabStripState.kt:
```diff
@@ -0,0 +1,69 @@
+    isPrivateMode: Boolean,
+): TabStripState {
+    return TabStripState(
+        tabs = tabs
+            .filter {
+                if (isPrivateMode) {
+                    it.content.private
+                } else {
+                    !it.content.private
+                }
+            }
```
Nice! 👍

rahulsainani commented:
Thanks @MozillaNoah for a very thorough review! As discussed offline, we want to iterate on it using the meta and create any follow ups there. If there are things we haven't addressed here, we will in the future.

rahulsainani commented:
Thanks everyone for testing it out, sharing the feedback and your excitement! It's great to see! I'll try to address more recent feedback here

– @Kuuh4 We'll look into it!
– @IrradiatedHaggis shortcuts would be great! it's something we'd like to add as well, it's outside the scope of this feature which still needs work. 
– @regs01 Some of the designs will update as we iterate, also the reason why this is behind secret settings right now. Tab strip now scrolling up with content is a known issue right now. 

🙌

systema-encephale commented:
I have tried to get the tab strip while using Firefox Nightly in the Samsung DeX desktop mode. I can see and activate the tab strip setting in the Secret Settings menu, but the strip itself does not appear. Is there something else I need to set?

rahulsainani commented:
@systema-encephale that's odd, were there any open tabs or did you open any new tabs? If so could you please share a video?

systema-encephale commented:
Ah, false alarm: I force-stopped Firefox and cleared the cache, and now the tabs are there. This is great, thank you for this addition!

rahulsainani commented:
@systema-encephale good to know! Thanks for the confirmation 🙌

NVins commented:
Please make it possible for the panel to be enabled on phones as well
Context menu when long pressing on a tab, and to have a mug animation when loading tabs, thanks!

NVins commented:
Please make it possible to enable the panel on phones as well.
The ability to move the tab bar down.
Context menu when long pressing on a tab.
And that when loading tabs, there was a circle animation, thanks!

Seddu commented:
> Please make it possible to enable the panel on phones as well. The ability to move the tab bar down. Context menu when long pressing on a tab. And that when loading tabs, there was a circle animation, thanks!

I second this. Brave and Vivaldi have it, don’t see an issue having the ability on firefox mobile as well. See images below.
![IMG_3585](https://github.com/mozilla-mobile/firefox-android/assets/152897099/7a281a83-5aa3-4173-9b01-06d7f8c55d12)
![IMG_3584](https://github.com/mozilla-mobile/firefox-android/assets/152897099/523de921-99e2-4c8f-bd55-22cf4705f72b)

marcel263 commented:
Dear Can it be force enabled based on screen size/resolution instead of detecting the form factor?

Currently sideloaded on Android TV box and option not showing up in the Secret Settings.

regs01 commented:
Another linked issue - address bar panel disappears, when opening find on page panel.

tkompis commented:
Tried installing again, downloaded the APK today following the steps above, I have no tab strip at all. Tried reinstalling too. Samsung S6 Lite. Do I have to change a setting somewhere? Or is it a bug?

rahulsainani commented:
@regs01 The address bar panal disappearing when opening find on page is an existing behaviour, or am i misunderstanding something here?

@NVins @Seddu @marcel263 @pigyee We hear you, we need some time to think these things through, Thanks a lot for the suggestions!

LostCursor commented:
An option to switch between normal tabs and this new tabbed view when folding and unfolding a folding phone, such as the Galaxy Fold 4 or Pixel Fold. Samsung Internet is capable of doing this, and so is Edge for Android.

Kuron85 commented:
There needs to be a tab animation when you click "+". It doesn't make sense to have a new tab button in the location it is in without that animation. It's confusing how the current action is and that "+" button and home button do the same action.

I hope this is able to be fixed and we can finally have new tab function like desktop Firefox and all other browsers.

Alosleon commented:
Hello! MiBox Android TV installs only the mobile version of Firefox Android Nightly
Please also add "Enable Tab Strip" in order to be visible and operative for the mobile version
Thanks

Kaoxt commented:
Is the New Tab button going to be addressed and have it open a tab directly? The Home button and New Tab button have the same function currently. Not being able to open a New Tab is maddening from the button itself.
